File: 05-binary_search/00033-Search_in_Rotated_Sorted_Array/00033-search_in_rotated_sorted_array.cpp
// 33. Search in Rotated Sorted Array
// https://leetcode.com/problems/search-in-rotated-sorted-array/


#include <iostream>
#include <vector>

using namespace std;


class Solution {
public:
    // SOLUTION
    int search(vector<int>& nums, int target) {
        int l = 0;
        int h = nums.size();

        while (l < h) {
            int m = (l+h) / 2;
            if (target < nums[0] < nums[m])
                l = m+1;
            else if (target >= nums[0] > nums[m])
                h = m;
            else if (nums[m] < target)
                l = m+1;
            else if (nums[m] > target)
                h = m;
            else
                return m;
        }

        return -1;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {4,5,6,7,0,1,2};
    int target = 0;

    // OUTPUT
    auto result = o.search(nums, target);
    cout << result << endl;

    return 0;
}

File: 05-binary_search/00033-Search_in_Rotated_Sorted_Array/00033-search_in_rotated_sorted_array.py
# 33. Search in Rotated Sorted Array
# https://leetcode.com/problems/search-in-rotated-sorted-array/


class Solution:
    # SOLUTION
    def search(self, nums: list[int], target: int) -> int:
        l, h = 0, len(nums)
        
        while l < h:
            m = (l+h) // 2
            if target < nums[0] < nums[m]:
                l = m+1
            elif target >= nums[0] > nums[m]:
                h = m
            elif nums[m] < target:
                l = m+1
            elif nums[m] > target:
                h = m
            else:
                return m

        return -1



if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums: list[int] = [4,5,6,7,0,1,2]
    target: int = 0

    # OUTPUT
    result = o.search(nums, target)
    print(result)
